### DISCUSSION TITLE
Pocketbase with Hooks

### DISCUSSION THREAD

**Discussion by siniradam** [pocketbase/pocketbase#2776] [Show and tell]
Hello folks,

I've created this with the help of a [friend](https://github.com/gurkanguray).
https://github.com/siniradam/PocketbaseHooks

It creates an additional table called `hooks`, if you add a record to this table containing `table`/collection name along with an `URL` and `op`eration name, it posts inserted/updated/deleted record to this endpoint URL with POST like `Hasura` does.

Unfortunately, I don't know how to write GO, so I mostly wrote it by searching and looking at other code pieces written online and thankfully my friend fixed my mistakes. Probably there are better ways to do some stuff in there, any critics and suggestions are welcome.

I wish to turn this into a plugin, I'm not sure if it's possible. 
Over and out. ☮️

One more thing: is it possible to define hooks in migration files?
Edit: Another thing, my ultimate goal is [PocketCloud](https://github.com/siniradam/PocketCloudFunctions), simply creating a portable BaaS with cloud functions.

**Comment by ganigeorgiev** [comment]
_Side note: You'll be able to attach to the [PocketBase event hooks](https://pocketbase.io/docs/event-hooks/) with JavaScript with the next v0.17.0 release. Some examples and info could be found in https://github.com/pocketbase/pocketbase/discussions/2767#discussioncomment-6245393. It is not a node/deno/etc. environment and loading external JS modules may not work as expected but you'll be able to do mostly the same things you can do with the Go hooks._

**Reply by benallfree** [reply]
This is cool!

**Reply by siniradam** [reply]
This is great, I've been meaning to do what I did for a while, but just couldn't find the time for it. I'm glad it is an upcoming feature now.
It's all I needed, wanted to simply add hooks anytime without changing much and triggering an external function when a `user` is created like Firebase Cloud Functions `exports.processSignUp = functions.auth.user().onCreate((user)=>{});`

**Reply by ganigeorgiev** [reply]
@siniradam _Note that the above is not http-based hooks, but application level hooks. It executes as wrapped Go code under the hood._

**Reply by webdeb** [reply]
Thats cool is more JS integration planned?

**Reply by ganigeorgiev** [reply]
@webdeb At this stage, probably No as the above should cover most of the cases (routing, database, emails, migrations, etc.). Depending on the user demand we may eventually add some additional Go->JS bindings for templates rendering and other similar utility helpers. Do you have anything particularly in mind?

**Reply by webdeb** [reply]
@ganigeorgiev I agree, this will cover a lot. And I did't yet started to use PB just looking around what it is, and it's very interesting so far. 👍

**Reply by siniradam** [reply]
@ganigeorgiev I was hoping for application-level hooks, along with HTTP, If I can manage I'll try to send commands via IPC or RPC. I'm still trying to figure out some low-level details about cloud functions, although [the thing I wrote](https://github.com/siniradam/PocketCloudFunctions) may not be the thing it is supposed to be but feels like it could be something useful.

If I can (if possible) would like to create a plugin, just a single line to import & execute in a pocket base build, that handles new records or events like user creation. 

In my code, it is just spawning express servers in cluster mode, but for a true cloud function, each function should be spawned & executed when it's needed. My approach might not be the best. Still, there is a lot to cover.

In the end, I can bundle it with Pocketbase via docker. A tiny BaaS. Right now I'm intrigued with making into a single file, with something like [pkg](https://github.com/vercel/pkg) If one day, Pocketbase would support plugins that would be a much better approach for me. Instead of making changes in Pocketbase and having a custom build.

**Reply by ganigeorgiev** [reply]
@siniradam 

> If I can (if possible) would like to create a plugin, just a single line to import & execute in a pocket base build, that handles new records or events like user creation.

You should be able to create a Go plugin by exporting a function/struct that accept [`core.App` interface](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.16.6/core#App) as argument. For an example you can check the `jsvm.MustRegister/jsvm.Register` methods in https://github.com/pocketbase/pocketbase/blob/develop/plugins/jsvm/jsvm.go#L62.
The drawback is that the above allow only backend plugin like behavior. The Admin UI at the moment is not extendable (_this will be handled in the future as part of #898_).

**Reply by siniradam** [reply]
This is more than enough right now. Just tapping into events would cover a huge base. I don't want to modify any default behavior in PB to keep compatibility at the highest.

**Reply by ganigeorgiev** [reply]
Just for info in case someone end up here - the js app hooks integration unfortunately will have to be deleted as it cannot work efficiently with how PocketBase hooks execute (for more details you can read in https://github.com/pocketbase/pocketbase/discussions/2825).

**Reply by ganigeorgiev** [reply]
_Second update: the js app hooks might be implemented with the next release with some limitations. For more details please check https://github.com/pocketbase/pocketbase/discussions/2825#discussioncomment-6376210._

**Comment by benallfree** [comment]
@siniradam If you would like to collaborate with https://pockethost.io, we are working on cloud functions as well.

**Reply by siniradam** [reply]
Hello Ben, @benallfree I would love that, I'll write you soon, kinda busy day.
